Add documentation on how to pass and fail tests (#2492)
cocotb/cocotb: documentation/source/writing_testbenches.rst

:meth:`~cocotb.handle.HierarchyObject._id` like in the following example:
``dut._id("_some_signal", extended=False)``.

> Passing And Failing Tests

Reviewer: ```suggestion
Passing and Failing Tests
```
(following https://en.wikipedia.org/wiki/Title_case)

=========================

A cocotb test is considered to have `failed` if the test coroutine,
or any coroutine :func:`~cocotb.fork`\ ed by the test coroutine,

> fails an ``assert`` statement or raises an :exc:`cocotb.result.TestFailure`.

Reviewer: ```suggestion
fails an ``assert`` statement or raises a :exc:`cocotb.result.TestFailure`.
```

Below are examples of `failing` tests.

.. code-block:: python3

    @cocotb.test()
    async def test(dut):
        assert 1 > 2, "Testing the obvious"

    @cocotb.test()
    async def test(dut):
        raise TestFailure("Reason")

    @cocotb.test()
    async def test(dut):
        async def fails_test():
            assert 1 > 2
        cocotb.fork(fails_test())
        await Timer(10, 'ns')

    @cocotb.test()
    async def test(dut):
        async def fails_test():
            raise TestFailure("Reason")
        cocotb.fork(fails_test())
        await Timer(10, 'ns')

When a test fails, a stacktrace is printed.

> If ``pytest`` is installed and ``assert`` statements are used,
> a more informative stacktrace, including the values that caused the ``assert`` to fail, is printed.

Reviewer: ```suggestion
If ``pytest`` is installed and ``assert`` statements are used,
a more informative stacktrace is printed which includes the values that caused the ``assert`` to fail.
```
